src/borg/helpers/time.py

```diff
@@ -185,3 +185,100 @@ def isoformat(self):
 def archive_ts_now():
     """return tz-aware datetime obj for current time for usage as archive timestamp"""
     return datetime.now(timezone.utc)  # utc time / utc timezone
+
+class DatePatternError(ValueError):
+    """Raised when a date: archive pattern cannot be parsed."""
+
+
+def local(dt: datetime) -> datetime:
+    """Attach the system local timezone to naive dt without converting."""
+    if dt.tzinfo is None:
+        dt = dt.replace(tzinfo=datetime.now().astimezone().tzinfo)
+    return dt
+
+
+def exact_predicate(dt: datetime):
+    """Return predicate matching archives whose ts equals dt (UTC)."""
+    dt_utc = local(dt).astimezone(timezone.utc)
+    return lambda ts: ts == dt_utc
+
+
+def interval_predicate(start: datetime, end: datetime):
+    start_utc = local(start).astimezone(timezone.utc)
+    end_utc   = local(end).astimezone(timezone.utc)
+    return lambda ts: start_utc <= ts < end_utc
+
+
+def compile_date_pattern(expr: str):
+    """
+    Turn a date: expression into a predicate ts->bool.
+    Supports:
+      1) Full ISO‑8601 timestamps with minute (and optional seconds/fraction)
+      2) Hour-only:   YYYY‑MM‑DDTHH      -> interval of 1 hour
+      3) Minute-only: YYYY‑MM‑DDTHH:MM   -> interval of 1 minute
+      4) YYYY, YYYY‑MM, YYYY‑MM‑DD       -> day/month/year intervals
+      5) Unix epoch (@123456789)         -> exact match
+    Naive inputs are assumed local, then converted into UTC.
+    TODO: verify working for fractional seconds; add timezone support.
+    """
+    expr = expr.strip()
+
+    # 1) Full timestamp (with fraction)
+    full_re = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+")
+    if full_re.match(expr):
+        dt = parse_local_timestamp(expr, tzinfo=timezone.utc)
+        return exact_predicate(dt) # no interval, since we have a fractional timestamp
+
+    # 2) Seconds-only
+    second_re = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}$")
+    if second_re.match(expr):
+        start = parse_local_timestamp(expr, tzinfo=timezone.utc)
+        return interval_predicate(start, start + timedelta(seconds=1))
+
+    # 3) Minute-only
+    minute_re = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$")
+    if minute_re.match(expr):
+        start = parse_local_timestamp(expr + ":00", tzinfo=timezone.utc)
+        return interval_predicate(start, start + timedelta(minutes=1))
+
+    # 4) Hour-only
+    hour_re = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}$")
+    if hour_re.match(expr):
+        start = parse_local_timestamp(expr + ":00:00", tzinfo=timezone.utc)
+        return interval_predicate(start, start + timedelta(hours=1))
```

Comment:
Maybe just use 1 regex with group names (`(?P<name>...)` that covers 1) .. 4) and also YYYY, YYYY-MM, YYYY-MM-DD cases from below.

After a single `m = re.match(regex, expr)`, you can check `m.groupdict()` in the right order (fraction, S, M, H, d, m, y) to determine which case you have.

Reply:
Maybe use re.VERBOSE so you can have a multi-line, commented regex for this.